### PR TITLE
Implement WalkJob RPC

### PR DESCRIPTION
**Pull request by FahadBSyed**
This is an implementation of the walk job RPC. Here are a few high level changes:
1. Protos for pjs.proto have been updated so a user can control the algorithm for walking. This will be useful for recursive waiting usecases
2. Generated code relating to the proto changes
3. Updates to pjsdb to support different walk file algorithms
4. Modifications to the api_server.go for pjs to support contexts
5. Implementation of the WalkJob RPC
6. A basic test that makes sure that walking the entire tree works using level order traversal.

### CODE REVIEW

File: src/internal/pjs/api_server.go [hunk 1]
```diff
@@ -42,7 +46,15 @@ func (a *apiServer) CreateJob(ctx context.Context, request *pjs.CreateJobRequest
 		}
 		inputs = append(inputs, fileset.PinnedFileset(*inputID))
 	}
+	var parent pjsdb.JobID
+	if request.Context != "" {
+		parent, err = a.resolveJobCtx(ctx, request.Context)
+		if err != nil {
+			return nil, errors.Wrap(err, "walk job")
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:53]: wrap the error with information that the caller of this function doesn't know, namely the name of the function that we're calling without the caller of CreateJob's knowledge; "resolveJobCtx".

File: src/internal/pjs/api_server.go [hunk 2]
```diff
@@ -156,7 +143,7 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
 			inputs = append(inputs, filesetID.HexString())
 		}
 		if err := srv.Send(&pjs.ProcessQueueResponse{
-			Context: "context is not implemented",
+			Context: hex.EncodeToString(jobCtx.Token),
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:146]: It would be nice if Token returned an object that had a String() method on it, so that someone doesn't decide here that it's hex.EncodeToString and someone in the future decides it's base64.EncodeToString.

File: src/internal/pjs/api_server.go [hunk 3]
```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
 		}
 	}
 }
+
+func (a *apiServer) WalkJob(req *pjs.WalkJobRequest, srv pjs.API_WalkJobServer) (err error) {
+	ctx := pctx.Child(srv.Context(), "walkJob")
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:181]: FWIW, srv.Context() is already named WalkJob (by grpc interceptors, when called thru grpc).

I like log.SpanContext for per-operation contexts.  You get a child context that you get to name, but it automatically reports success/failure, how long the deadline on the context is (or its error if it's done), etc.

File: src/internal/pjs/api_server.go [hunk 4]
```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
...
+		id = reqID // defaulting to this for testing until auth is implemented.
+	}
+
+	// walk and stream back results.
+	var jobs []pjsdb.Job
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		jobs, err = pjsdb.WalkJob(ctx, sqlTx, id, pjsdb.WalkAlgorithm(req.Algorithm.Number()))
+		return errors.Wrap(err, "with tx")
+	}); err != nil {
+		return errors.Wrap(err, "walk job")
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:206]: This is the with tx error, and the one above is the walk job error.

File: src/internal/pjs/api_server.go [hunk 5]
```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
...
+		}
+		resp := &pjs.ListJobResponse{
+			Id:   jobInfo.Job,
+			Info: jobInfo,
+			Details: &pjs.JobInfoDetails{
+				JobInfo: jobInfo,
+			},
+		}
+		if err := srv.Send(resp); err != nil {
+			return errors.Wrap(err, "walk job")
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:221]: wrap with `send`.  The caller of this function should wrap the error with `WalkJob`, not this function itself.  Only wrap errors with information the caller can't know.

In this case, you should Wrapf with `job` here.  That way, when Send fails we know how deep into `jobs` we are.

File: src/internal/pjs/api_server.go [hunk 6]
```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
...
+	var jobCtxDecoded []byte
+	jobCtxDecoded, err = hex.DecodeString(jobCtx)
+	if err != nil {
+		return 0, errors.Wrap(err, "walk job")
+	}
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		id, err = pjsdb.ResolveJobContext(ctx, sqlTx, jobCtxDecoded)
+		if err != nil {
+			// TODO(PJS): handle NotFound with a special error type that returns the right proto error code.
+			return errors.Wrap(err, "with tx")
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:237]: with tx / walk job inverted.

File: src/internal/pjs/api_server.go [hunk 7]
```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
...
+	for _, filesetID := range job.Inputs {
+		jobInfo.Input = append(jobInfo.Input, filesetID.HexString())
+	}
+	switch {
+	case !job.Queued.IsZero() && job.Processing.IsZero() && job.Done.IsZero():
+		jobInfo.State = pjs.JobState_QUEUED
+	case !job.Queued.IsZero() && !job.Processing.IsZero() && job.Done.IsZero():
+		jobInfo.State = pjs.JobState_PROCESSING
+	case !job.Queued.IsZero() && !job.Processing.IsZero() && !job.Done.IsZero():
+		jobInfo.State = pjs.JobState_DONE
```
Comment (jrockway) [inline src/internal/pjs/api_server.go:265]: These cases are kind of hard to follow.  It would be nice to name each one outside the switch statement `queued := !job.Queued.IsZero() && job.Processing.IsZero() && job.Done.IsZero()`, or even a method on job that returns the state enum.

File: src/internal/pjsdb/job_crud.go [hunk 8]
```diff
@@ -171,19 +171,34 @@ func CancelJob(ctx context.Context, tx *pachsql.Tx, id JobID) ([]JobID, error) {
...
 }
 
+// WalkAlgorithm is an enumerator for walking algorithms.
+type WalkAlgorithm int32
+
+const (
+	LevelOrder        = WalkAlgorithm(0)
+	PreOrder          = WalkAlgorithm(iota)
+	MirroredPostOrder = WalkAlgorithm(iota)
+)
```
Comment (jrockway) [inline src/internal/pjsdb/job_crud.go:181]:
```go
const (
    LevelOrder WalkAlgorithm = iota
    PreOrder
    MirroredPostOrder
)
```

File: src/internal/pjsdb/job_crud.go [hunk 9]
```diff
@@ -171,19 +171,34 @@ func CancelJob(ctx context.Context, tx *pachsql.Tx, id JobID) ([]JobID, error) {
...
 	}
-	records := make([]jobRecord, 0)
-	if err = sqlx.SelectContext(ctx, tx, &records, recursiveTraverseChildren+selectJobRecordPrefix+`
-		INNER JOIN children c ON j.id = c.id
-		GROUP BY j.id
-		ORDER BY MIN(depth) ASC;
-	`, job.ID); err != nil {
+	var records []jobRecord
+	var walker func(ctx context.Context, tx *pachsql.Tx, id JobID) ([]jobRecord, error)
+	switch algo {
```
Comment (jrockway) [inline src/internal/pjsdb/job_crud.go:192]: a comment `//exhaustive:enforce` will make the linter fail if you add a new WalkAlgorithm and don't update this switch statement.

File: src/internal/pjsdb/job_crud.go [hunk 10]
```diff
@@ -197,6 +212,48 @@ func WalkJob(ctx context.Context, tx *pachsql.Tx, id JobID) ([]Job, error) {
 	return jobs, nil
 }
 
+func walkLevelOrder(ctx context.Context, tx *pachsql.Tx, id JobID) (records []jobRecord, err error) {
+	records = make([]jobRecord, 0)
```
Comment (jrockway) [inline src/internal/pjsdb/job_crud.go:216]: `var records []jobRecord` is preferred over this declaration syntax: https://go.dev/wiki/CodeReviewComments#declaring-empty-slices

File: src/pjs/pjs.proto [hunk 11]
```diff
@@ -192,11 +192,30 @@ message ListJobResponse {
   JobInfoDetails details = 3;
 }
 
+enum WalkAlgorithm {
```
Comment (jrockway) [inline src/pjs/pjs.proto:195]: Good docs.  Document on the line above each entry so they work their way into the API docs.  (proto-docs.md in your PR.)

File: src/pjs/pjs.proto [hunk 12]
```diff
@@ -192,11 +192,30 @@ message ListJobResponse {
...
+  LEVEL_ORDER = 0; // similar to BFS, but specifically for trees. Traverse nodes in height order.
```
Comment (jrockway) [inline src/pjs/pjs.proto:196]: Something that happens with protos is when a client has a 4th algorithm added and the server doesn't have that algorithm (compiled against older protos), it will be mapped to the first entry here.  So LEVEL_ORDER will be silently selected.

The way to avoid this is to call the first entry UNKNOWN and then never use it.  If you get a request where the algorithm is UNKNOWN, you'll know that something is out of sync.

**Review by jrockway — APPROVED** [review]
Looks good.  Some nits.

File: src/internal/pjs/api_server.go (hunk 7 above)
Comment (FahadBSyed) [reply src/internal/pjs/api_server.go:265]: I just copied and pasted this code but I think we can just do:
```
case job.Done != time.Time{}:
case job.Processing != time.Time{}:
default
```

File: src/internal/pjsdb/job_crud.go (hunk 10 above)
Comment (FahadBSyed) [reply src/internal/pjsdb/job_crud.go:216]: I think I did this because I had wrongly assumed that SelectContext required the array to be already initialized. However I must have hallucinated that.

File: src/pjs/pjs.proto (hunk 12 above)
Comment (FahadBSyed) [reply src/pjs/pjs.proto:196]: Now I know why we always use UNKOWN in enums.

File: src/internal/pjsdb/job_crud.go (hunk 10 above)
Comment (jrockway) [reply src/internal/pjsdb/job_crud.go:216]: These two forms (`var x []T` and `x := make([]T, 0)`) are basically equivalent.  (Slice with len = 0, cap = 0.)  If you were passing this to (io.Writer).Read, then it's wanting you to set the length of the read by setting the length of the slice and `make([]byte, 42)` would be appropriate for that.

File: src/internal/pjs/api_server.go (hunk 2 above)
Comment (FahadBSyed) [reply src/internal/pjs/api_server.go:146]: I think the String() function actually prints the hash to avoid leaking the context token.

File: src/internal/pjsdb/job_crud.go (hunk 10 above)
Comment (jrockway) [reply src/internal/pjsdb/job_crud.go:216]: Oh and when it's the return value like this you don't even need `var`.  It's ready to append to.

File: src/internal/pjs/api_server.go (hunk 2 above)
Comment (jrockway) [reply src/internal/pjs/api_server.go:146]: I see.  That's fine.  In the event that someone does `log.Proto(<job>)` and you don't want the token to be logged, be sure to annotate it in the proto file with `[(log.mask) = true];`.  (There is also (log.half) if you want to only log half the string.  We use that for OIDC tokens.)

Comment (FahadBSyed) [reply src/internal/pjs/api_server.go:146]: Oh yeah, good idea!

File: src/internal/pjs/api_server.go [hunk 13]
```diff
@@ -184,8 +171,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
...
+		}
+		if err := srv.Send(resp); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("send, iteration=%d/%d", i, len(jobs)))
+		}
+	}
+	return nil
+}
+
+func (a *apiServer) resolveJobCtx(ctx context.Context, jobCtx string) (id pjsdb.JobID, err error) {
+	token, err := pjsdb.JobContextTokenFromString(jobCtx)
```
Comment (Zhang-Muyang) [inline src/internal/pjs/api_server.go:229]: I took a look at the failed bazel test because I want to reuse your fullBinaryTree and resolveContext method 😄.
I think the problem comes from this line. If you replace JobContextTokenFromString with JobContextTokenFromHex, that failed test will pass. This is because in api_server line 146, the token is already encoded into hex format.

Comment (FahadBSyed) [reply src/internal/pjs/api_server.go]: Thanks a bunch Muyang!
